src/opencl_common.c

```diff
@@ -2829,6 +2881,269 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
+	(*list)->count = 0;
+}
+
+static void buf_list_add_link(struct opencl_buffers *list, struct buffer_opencl *entry) {
+	entry->next = NULL;
+
+	if (list->tail)
+		list->tail = list->tail->next = entry;
+	else
+		list->tail = list->head = entry;
+
+	list->count++;
+}
+
+static void buf_list_add(struct opencl_buffers *list, cl_mem *data) {
+	struct buffer_opencl *entry;
+
+	entry = mem_alloc(sizeof(struct buffer_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
+
+	buf_list_add_link(list, entry);
+}
+
+static void buf_list_remove(struct opencl_buffers *list, cl_mem data) {
+	struct buffer_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
+			if (prev)
+				prev->next = current->next;
+
+			if (list->head == current)
+				list->head = current->next;
+			if (list->tail == current)
+				list->tail = prev;
+
+			*(current->variable) = NULL; //Is it really safe?
+			MEM_FREE(current);
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+static void krn_list_init(struct opencl_kernels **list) {
+	*list = mem_alloc(sizeof(struct opencl_kernels));
+	(*list)->tail = (*list)->head = NULL;
+	(*list)->count = 0;
+}
+
+static void krn_list_add_link(struct opencl_kernels *list, struct kernel_opencl *entry) {
+	entry->next = NULL;
+
+	if (list->tail)
+		list->tail = list->tail->next = entry;
+	else
+		list->tail = list->head = entry;
+
+	list->count++;
+}
+
+static void krn_list_add(struct opencl_kernels *list, cl_kernel *data) {
+	struct kernel_opencl *entry;
+
+	entry = mem_alloc(sizeof(struct kernel_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
+
+	krn_list_add_link(list, entry);
+}
+
+static void krn_list_remove(struct opencl_kernels *list, cl_kernel data) {
+	struct kernel_opencl *current, *prev = NULL;
+
+	for (current = list->head; current; current = current->next) {
+		if (current->data == data) {
+			if (prev)
+				prev->next = current->next;
+
+			if (list->head == current)
+				list->head = current->next;
+			if (list->tail == current)
+				list->tail = prev;
+
+			*(current->variable) = NULL; //Is it really safe?
+			MEM_FREE(current);
+
+			list->count--;
+
+			/* Done! */
+			break;
+		}
+		prev = current;
+	}
+}
+
+/*
+ * Use this function to alloc OpenCL buffers.
+ * Later, all buffers can be released at once using opencl_release_buf().
+ *
+ * Setting keep_track to zero (0) causes the function to NOT save a reference
+ * of the newly created buffer. Therefore, the buffer will not be released
+ * automatically by opencl_release_buf().
+ *
+ * IMPORTANT: the function keeps a referent to _buffer. So, it can't be freed
+ *             before a call to opencl_release_buf();
+ */
+cl_mem opencl_create_buf(cl_mem *_buffer, cl_mem_flags _flags, size_t _size,
+            int keep_track) {
+	cl_int __errCreateBuffer;
+
+	if (*(_buffer))
+		opencl_release_buf(_buffer);
```

Comment:
We could regard this as an error instead. Or, we could make that depend on `-DDEBUG`

Reply:
Done